### PR TITLE
tree2: Better and more consistent field cursor support

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

This better formalizes the convention of using above root placeholder nodes, leverages this to provide field cursor builders, and makes cursor related function naming more consistent.

## Breaking Changes

Some cursor related functions have been renamed and had argument order changed, but nothing was removed.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: experimental/dds/tree2/src/core/tree/types.ts [hunk 1]
```diff
@@ -161,3 +161,10 @@ export interface NodeData {
 	 */
 	readonly type: TreeNodeSchemaIdentifier;
 }
+
+/**
+ * When creating a node which sits above the detached fields, and thus is not a real node and who's type should not matter, use this type to indicate that.
```
Comment (Josmithr) [inline experimental/dds/tree2/src/core/tree/types.ts:166]: Nit (phrasing is a bit confusing). Maybe something like:
```suggestion
 * Use this type to indicate that a node sits above the detached fields, and thus is not a real node and who's type should not matter.
```

File: experimental/dds/tree2/src/feature-libraries/treeCursorUtils.ts [hunk 2]
```diff
@@ -43,21 +46,36 @@ export interface CursorWithNode<TNode> extends ITreeCursorSynchronous {
...
+): CursorWithNode<TNode> {
+	const cursor = stackTreeNodeCursor(adapter, root);
+	// Because the root node in `stackTreeNodeCursor` is treated as the above detached fields node,
+	// using it then just entering the correct field doesn't mess up the paths reported by the cursor.
+	cursor.enterField(detachedFieldAsKey(detachedField));
+	return cursor;
+}
+
+/**
+ * Provides functionality to allow a {@link stackTreeNodeCursor} to implement a cursor.
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/treeCursorUtils.ts:78]: It's intended to be more general than this single use-case now, right? Should we update the docs to be a bit more semantic?

File: experimental/dds/tree2/src/feature-libraries/treeTextCursor.ts [hunk 3]
```diff
@@ -45,14 +49,33 @@ import { CursorAdapter, singleStackTreeCursor } from "./treeCursorUtils";
 /**
  * Create a cursor, in `nodes` mode at the root of the provided tree.
  *
- * @returns an {@link ITreeCursorSynchronous} for a single {@link JsonableTree}.
+ * @returns an {@link ITreeCursorSynchronous} in nodes mode for a single {@link JsonableTree}.
+ * @remarks
+ * Do not confuse this with {@link JsonableTree} with the JSON domain:
+ * this takes in data in a specific format that is json compatible (except for FluidHandel values).
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/treeTextCursor.ts:55]:
```suggestion
 * this takes in data in a specific format that is json compatible (except for FluidHandle values).
```

Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/treeTextCursor.ts:55]: What do we do in the FluidHandle case? Is that still TBD?

**Review by Josmithr — APPROVED** [review]
Left a couple of documentation nitpicks. Otherwise looks good to me!

File: experimental/dds/tree2/src/feature-libraries/treeTextCursor.ts (hunk 3 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/treeTextCursor.ts:55]: For fluid handles, we include IFluidHandles (which is clear from the JsonableTreeType if you look at what it allows). This is simply called out here as they are not actually json compatible, so the name is a bit inaccurate.

File: experimental/dds/tree2/src/feature-libraries/treeCursorUtils.ts (hunk 2 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/treeCursorUtils.ts:78]:
```suggestion
 * Provides functionality to allow a {@link stackTreeNodeCursor} and {@link stackTreeFieldCursor} to implement cursors.
```

**Pull request merged**